### PR TITLE
Speed up NTM reader

### PR DESCRIPTION
**Pull request by reyoung**
* Load Data --> 64 sec
* Shuffle/Batch --> 14 sec

default
[train.baseline.txt](https://github.com/PaddlePaddle/models/files/2288974/train.baseline.txt)
[train.pylogger.txt](https://github.com/PaddlePaddle/models/files/2288975/train.pylogger.txt)

### CODE REVIEW

File: fluid/neural_machine_translation/transformer/reader.py [hunk 1]
```diff
@@ -263,95 +262,41 @@ def load_dict(dict_path, reverse=False):
...
-            batch_data = []
-            max_len = -1
-            batch_max_seq_len = -1
-
-            while True:
-                sample = pool.next(look=True)
+            infos = self._sample_infos
+            self._random.shuffle(infos)
+        else:
+            infos = self._sample_infos
```
Comment (guoshengCS) [inline fluid/neural_machine_translation/transformer/reader.py]: We may also need local sort to sort by groups as SortType.POOL indicates. It is better to support this.

File: fluid/neural_machine_translation/transformer/reader.py [hunk 2]
```diff
@@ -263,95 +262,41 @@ def load_dict(dict_path, reverse=False):
...
         if self._sort_type == SortType.GLOBAL:
-            if not self._sorted:
-                self._sample_idxs.sort(
-                    key=lambda idx: max(len(self._src_seq_ids[idx]),
-                    len(self._trg_seq_ids[idx] if not self._only_src else 0))
-                )
-                self._sorted = True
+            infos = sorted(
+                self._sample_infos,
+                key=lambda x: max(x[1], x[2]) if not self._only_src else x[1])
```
Comment (guoshengCS) [inline fluid/neural_machine_translation/transformer/reader.py]: Do we need states to keep the sort results and batches, and thus can avoid extra processing and batch generation

**Review by guoshengCS — COMMENTED** [review]
Thanks for this work!

File: fluid/neural_machine_translation/transformer/reader.py [hunk 3]
```diff
@@ -263,95 +262,41 @@ def load_dict(dict_path, reverse=False):
...
+                batches.append([info.i for info in batch])
 
-                max_len = max(max_len, len(sample[0]))
+        if not self._clip_last_batch and len(batch_creator.batch) != 0:
+            batches.append([info.i for info in batch_creator.batch])
 
-                if not self._only_src:
-                    max_len = max(max_len, len(sample[1]))
+        if self._shuffle:
+            self._random.shuffle(batches)
```
Comment (guoshengCS) [inline fluid/neural_machine_translation/transformer/reader.py]: Should it be the `self._shuffle_batch` flag

**Review by guoshengCS — APPROVED** [review]
LGTM

**Comment by reyoung** [comment]
|  | Paddle Version | Models Version|
| --- | --- | --- | 
| Baseline |  66be53264e |  be3dcb4e30 |
| This PR | 66be53264e | be3dcb4e30 |

**Pull request merged**